internal/service/agent/handler.go

```diff
@@ -0,0 +1,121 @@
+package service
+
+import (
+	"context"
+	"errors"
+
+	"github.com/flightctl/flightctl/internal/api/server"
+	agentServer "github.com/flightctl/flightctl/internal/api/server/agent"
+	"github.com/flightctl/flightctl/internal/crypto"
+	"github.com/flightctl/flightctl/internal/server/tlscontext"
+	"github.com/flightctl/flightctl/internal/service/common"
+	"github.com/flightctl/flightctl/internal/store"
+	"github.com/sirupsen/logrus"
+)
+
+type AgentServiceHandler struct {
+	store store.Store
+	ca    *crypto.CA
+	log   logrus.FieldLogger
+}
+
+// Make sure we conform to servers Service interface
+var _ agentServer.Service = (*AgentServiceHandler)(nil)
+
+func ValidateDeviceAccessFromContext(ctx context.Context, name string, log logrus.FieldLogger) error {
+	cn, ok := ctx.Value(tlscontext.TLSCommonNameContextKey).(string)
+	if !ok {
+		log.Warningf("an attempt to access device %q without a CN in tls certificate has been detected", name)
+		return errors.New("no common name in certificate")
+	}
+	if expectedCn, err := crypto.CNFromDeviceName(name); err != nil {
+		return err
+	} else if cn != expectedCn {
+		log.Warningf("an attempt to access device %q with a certificate with CN %q has been detected", name, cn)
+		return errors.New("invalid tls CN for device")
+	}
```

Comment:
optional: IMO easier to read
```suggestion
        expectedCn, err := crypto.CNFromDeviceName(name)
	if err != nil {
		return err
	}
	if cn != expectedCn {
		log.Warningf("an attempt to access device %q with a certificate with CN %q has been detected", name, cn)
		return errors.New("invalid tls CN for device")
	}
```